File: go/bob/bob.go
// This is bob, a lackadaisical teenager.
package bob

import (
	"strings"
	"unicode"
)

// Hey tells bob to answer your remark.
func Hey(remark string) string {
	remark = trim(remark)

	answer := "Whatever."

	isQuestion := strings.HasSuffix(remark, "?")
	isYelled := strings.ToUpper(remark) == remark && strings.ContainsAny(strings.ToLower(remark), "abcdefghijklmnopqrstuvwxyz")

	if isQuestion {
		answer = "Sure."
	}
	if isYelled {
		answer = "Whoa, chill out!"
	}
	if isYelled && isQuestion {
		answer = "Calm down, I know what I'm doing!"
	}
	if remark == "" {
		answer = "Fine. Be that way!"
	}

	return answer
}

func trim(remark string) string {
	return strings.TrimFunc(remark, func(r rune) bool {
		return !(unicode.IsDigit(r) || unicode.IsLetter(r) || strings.ContainsAny(string(r), "();:-!§$%&/()=?)"))
	})
}

File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(raw_statement: String): String = {
    val statement = raw_statement.trim()
    val isQuestion = statement.endsWith("?")
    val containsLetters = statement.exists(_.isLetter)
    val isYelled = statement.toUpperCase == statement && containsLetters

    if (statement.isBlank) "Fine. Be that way!"
    else if (isYelled && isQuestion) "Calm down, I know what I'm doing!"
    else if (isYelled) "Whoa, chill out!"
    else if (isQuestion) "Sure."
    else "Whatever."
  }
}
